source/azure_iot_provisioning_client.c

```diff
@@ -0,0 +1,963 @@
+            return;
+        }
+
+        ( void ) memset( ( void * ) &xMQTTPublishInfo, 0x00, sizeof( xMQTTPublishInfo ) );
+        xMQTTPublishInfo.xQOS = eAzureIoTMQTTQoS0;
+        xMQTTPublishInfo.pcTopicName = xAzureIoTProvisioningClientHandle->_internal.pucProvisioningScratchBuffer;
+        xMQTTPublishInfo.usTopicNameLength = ( uint16_t ) xMQTTTopicLength;
+        xMQTTPublishInfo.pvPayload = xAzureIoTProvisioningClientHandle->_internal.pucProvisioningScratchBuffer + xMQTTTopicLength;
+        xMQTTPublishInfo.xPayloadLength = xMQTTPayloadLength;
+        usPublishPacketIdentifier = AzureIoTMQTT_GetPacketId( &( xAzureIoTProvisioningClientHandle->_internal.xMQTTContext ) );
+
+        if( ( xResult = AzureIoTMQTT_Publish( &( xAzureIoTProvisioningClientHandle->_internal.xMQTTContext ),
+                                              &xMQTTPublishInfo, usPublishPacketIdentifier ) ) != eAzureIoTMQTTSuccess )
+        {
+            AZLogError( ( "Failed to publish. Error=%d", xResult ) );
+            xReturn = eAzureIoTProvisioningPublishFailed;
+        }
+        else
+        {
+            xReturn = eAzureIoTProvisioningSuccess;
+        }
+
+        prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, xReturn );
+    }
+    else
+    {
+        AZLogWarn( ( "Request action called in wrong state %d\r\n",
+                     xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState ) );
+    }
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Implementation of parsing response action, this action is only allowed in azureiotprovisioningWF_STATE_RESPONSE
+ *
+ * */
+static void prvAzureIoTProvisioningClientParseResponse( AzureIoTProvisioningClientHandle_t xAzureIoTProvisioningClientHandle )
+{
+    az_result xCoreResult;
+    az_span xPayload = az_span_create( xAzureIoTProvisioningClientHandle->_internal.pucProvisioningLastResponsePayload,
+                                       ( int32_t ) xAzureIoTProvisioningClientHandle->_internal.xProvisioningLastResponsePayloadLength );
+    az_span xTopic = az_span_create( xAzureIoTProvisioningClientHandle->_internal.pucProvisioningLastResponseTopic,
+                                     ( int32_t ) xAzureIoTProvisioningClientHandle->_internal.usProvisioningLastResponseTopicLength );
+
+    /* Check the state.  */
+    if( xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState == azureiotprovisioningWF_STATE_RESPONSE )
+    {
+        xCoreResult =
+            az_iot_provisioning_client_parse_received_topic_and_payload( &xAzureIoTProvisioningClientHandle->_internal.xProvisioningClientCore,
+                                                                         xTopic, xPayload, &xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse );
+
+        if( az_result_failed( xCoreResult ) )
+        {
+            prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, eAzureIoTProvisioningFailed );
```

Comment:
Nit  - here you call the UpdateState function and then log error but a few lines later you call the log error and then the update state in different order.

I'd recommend logging 1st though I don't care that much beyond consistency.  The "Log/DosomeError" is how I've seen this historically, and on the case that something weird happens in the function called in the else{errorBlock}, then there's at least a chance that the log error will be output.  (Pretending stdout gets flushed out of course:)).